Gabi/atqr 125 create contacts value objects (Jira: ATQR-78 )

GabrielaRsanches opened the pull request:
implemented Email, Contact and Invitee Value Objects following disl structure , using class-validator to validate their proprieties when necessary. 
I created a validateOrReject.ts file in services folder containing this validateOrReject function in a way that it can be used in other contexts. 
I tried to make a rebase before creating this pull request, but my branch is up to date

Commit:
perf(contacts.disl and email.interface.ts): emailAddress inteface, OV Email on disl and domain

interface for EmailAddress now exists and has validation test. Concept EmailAddress is in
contacts.disl

Commit:
chore(parcel was updated to its right version): parcel is now on version 2.8.2

Commit:
chore(package.json): script "atqr:start:test" for testing with jest

Commit:
perf(contact, invitee and email value objects): implementation, testing and refactor

contact, invitee and email value objects are implemented

Commit:
feat(service validateorreject.ts): class-validator function "validateOrReject" is reusable

function is in services folder to be used in other files

Commit:
refactor(contact, email and invitee value objects): classes Value Object are extending ValueObject

Contact, Email and Invitee are extending ValueObject class from ValueObject.ts

yfernandes commented on libs/atqr/domain/src/lib/domain-spec/concept-set/Contacts.disl:
```diff
@@ -13,5 +13,10 @@ domain ATQR
     @email: DISL.EmailAddress
     @role: ParticipationRole
 
-  concept Invitee expands on Contact
```
Correct syntax for Disl definition is to use expands. Extends is TS syntax

yfernandes commented on libs/atqr/domain/src/lib/domain-spec/concept-set/Contacts.disl:
```diff
@@ -3,7 +3,7 @@ domain ATQR
   //--------------------------------//
   // Contacts
   //--------------------------------//
-  concept Email
```
This change is semantically and syntatically incorrect!

GabrielaRsanches replied:
don't really understand . I'm implementing Email, Contact and Invitee VO. Email wasn't implemented on disl , so I followed the syntax I saw in the other two already implemented

GabrielaRsanches replied:
@yfernandes

GabrielaRsanches replied:
@yfernandes

yfernandes replied:
It's syntatically wrong because Disl syntax is to use expands instead of extends.

yfernandes commented on libs/atqr/domain/src/lib/services/validateOrReject.ts:
```diff
@@ -0,0 +1,11 @@
+import { validateOrReject } from 'class-validator';
```
This function is not a service!

yfernandes commented on package.json:
```diff
@@ -59,12 +60,14 @@
     "node-ts": "5.1.2",
     "nodemailer": "6.7.3",
     "nodemailer-mailgun-transport": "2.1.3",
+    "parcel": "^2.8.2",
```
The use of wildcard selectors is disallowed, and has been discussed many times.

yfernandes commented on package.json:
```diff
@@ -59,12 +60,14 @@
     "pug-lexer": "5.0.1",
     "reflect-metadata": "0.1.13",
     "regenerator-runtime": "0.13.7",
+    "regex": "^0.1.1",
```
Unnecessary dependency

yfernandes commented on package.json:
```diff
@@ -90,8 +93,8 @@
     "@nrwl/workspace": "13.4.5",
-    "@parcel/transformer-pug": "2.8.0",
-    "@parcel/transformer-sass": "2.8.0",
+    "@parcel/transformer-pug": "^2.8.2",
```
Wildcard selector.

yfernandes commented on package.json:
```diff
@@ -90,8 +93,8 @@
-    "@parcel/transformer-pug": "2.8.0",
-    "@parcel/transformer-sass": "2.8.0",
+    "@parcel/transformer-pug": "^2.8.2",
+    "@parcel/transformer-sass": "^2.8.2",
```
Wildcard selector.

yfernandes commented on libs/atqr/domain/src/lib/EmailService/email.interface.ts:
```diff
@@ -1,19 +1,31 @@
-    public message?: string
+    public body?: string
   ) {
+    super(Email, Email[to.toString()][subject][body]);
```
Incorrect call to super. The types passed are wrong

yfernandes commented on libs/atqr/domain/src/lib/EmailService/email.interface.ts:
```diff
@@ -1,19 +1,31 @@
     this.subject = subject;
-    this.message = message;
+    this.body = body;
+    return new Email(
```
Recursive constructor call.

yfernandes commented on libs/atqr/domain/src/lib/EmailService/email.interface.ts:
```diff
@@ -28,7 +40,8 @@ export class Email implements IEmail {
+    return this.body;
   }
 }
+validate(Email); //uses validateOrReject from class-validator
```
Dangling function, and function called with incorrect parameters.

yfernandes commented on libs/atqr/domain/src/lib/services/validateOrReject.ts:
```diff
@@ -0,0 +1,11 @@
+import { validateOrReject } from 'class-validator';
+export async function validate(props) {
```
Function has implicity any

yfernandes commented on libs/atqr/domain/src/lib/contact/contact.valueObject.ts:
```diff
@@ -0,0 +1,30 @@
+    );
+  }
+}
+validate(Contact); //uses validateOrReject from class-validator
```
Dangling function, and function called with incorrect parameters.

yfernandes commented on libs/atqr/domain/src/lib/EmailService/email.interface.ts:
```diff
@@ -1,19 +1,31 @@
     public to: EmailAddress,
     public subject: string,
-    public message?: string
+    public body?: string
```
Body is not optional.

yfernandes reviewed (commented):
Besides the comments in code, i would like to know why was the email.service deleted.

There were a lot of misused types in this commit, pay more attention in the future, an absolutely in no case allow implicity any

yfernandes commented:
Please resolve all comments and re-request a review.

Commit:
refactor(package.json): wildcard selectors were deleted

Commit:
refactor(vo contact and email): vO are instantiating new object with proprieties updated

super call is receving right params

GabrielaRsanches commented:
> 

email.service was deleted in a group agreement , because it had the same content as email.interface and I worked on email.interface to add new structure according to .disl 
So email.service had no propose . Email service functions are been implemented in notification.service. Please check out, because I pointed this to you and had no answer 
@yfernandes

albnunes committed:
(ATQR-128) Create notifications Templatedatatypes (#73)

* docs: notification documentation has been refactored to be more explanatory

* feat: the dependency class-validator was added in the json package

* feat: dto created to contain the data of the message that will be sent

* feat: implementation of message templates that can be sent

* feat: the tempalte data type is implemented correctly

raquellenn committed:
(ATQR-123) Updates enums to match concept set (#76)

* chore(enums from the concept-set disl): updated the enums to match the definition in disl

* fix(domain/challenge): removes Duplicate enums from domain challenge

* refactor(domain): moves misplaced concept-set enums from Web app to Domain

* refactor(web/services): removes duplicated enums and updates indexes

* fix(challenge.entity): changes supervisor status to participation status

* fix(web/service/interface): removes payment enum from interfaces

Co-authored-by: Yago Fernandes <yagoalmeida@gmail.com>

rairezende committed:
(ATQR-143) Corrects the spelling of the "aggregates" folder (#77)

✅ Closes: #143

Commit:
refactor: ready to another review

rairezende committed:
(ATQR-145) Implemented Payments enums (#87)

* feat: 🎸 Issue #145 -  Implemented Payments enums

✅ Closes: 145

* fix(domain/paymentmethod): exports PaymentMethod enums

Co-authored-by: Yago Fernandes <yagoalmeida@gmail.com>

svallory commented on package.json:
```diff
@@ -9,6 +9,7 @@
     "atqr:start:api": "nx serve api",
     "postatqr:start:api": "docker compose down",
     "atqr:start:web": "yarn parcel apps/atqr/web/src/pages/**/*.pug --open",
+    "atqr:start:test": "yarn run jest",
```
"start" is generally used to start the application. Rename the task to simply "atqr:test"

svallory commented on libs/atqr/domain/src/lib/invitee/invitee.valueObject.ts:
```diff
@@ -0,0 +1,28 @@
+interface IInvitee extends Contact {
+  status: ParticipationStatus;
+}
+class Invitee extends ValueObject<Invitee> implements IInvitee {
```
Looks like you forgot to refactor `Invitee` as you did for `Contact`

svallory commented on libs/atqr/domain/src/lib/EmailService/email.interface.ts:
```diff
@@ -1,34 +1,41 @@
+  @IsEmail()
+  emailAddress: string;
+}
 export interface IEmail {
```
We don't need an email interface. Remove it and make sure it is not used anywhere

svallory commented on libs/atqr/domain/src/lib/EmailService/email.interface.ts:
```diff
@@ -1,34 +1,41 @@
   constructor(
-    public to: EmailAddress,
+    public readonly to: EmailAddress,
     public subject: string,
```
subject and body properties must also be `readonly`

svallory commented on libs/atqr/domain/src/lib/contact/contact.valueObject.ts:
```diff
@@ -0,0 +1,38 @@
+  email: EmailAddress;
+  role: ParticipationRole;
+
+  constructor(name: string, email: EmailAddress, role: ParticipationRole) {
```
Use `public readonly prop: type` and you can remove property declarations

GabrielaRsanches replied:
@svallory I did it because decorators can not be used inside constructor param . Looks better with all proprieties declared than only name and its decorators declared and the other two just public readonly inside of the constructor params , right ?

svallory commented on libs/atqr/domain/src/lib/contact/contact.valueObject.ts:
```diff
@@ -0,0 +1,38 @@
+export class Contact extends ValueObject<Contact> {
+  @MinLength(5)
+  @MaxLength(20)
+  name: string;
```
all properties in value objects must be read-only

svallory commented on libs/atqr/domain/src/lib/contact/contact.valueObject.ts:
```diff
@@ -0,0 +1,38 @@
+
+  public setName(newName: string): Contact {
+    return this.newInstanceWith({
+      name: newName,
```
If you make the name of the parameter the same as the name of the property you can wite a shorter object. In this case, if you rename `newName` to `name`, you can simply write: 

`this.newInstanceWith({ name });`

And it will be the same as writing

`this.newInstanceWith({ name: name });`

svallory reviewed (commented):
Let's create a `valueObjects` folder in lib and move all the value object files there. Also, please remove `.valueObject` and `.interface` from file names.

Rename `email.service.ts` to `EmailService.ts`

rairezende committed:
(ATQR-124) Implemented Enrollment and Officiation Object Values (#75)

* feat: 🎸 implemented Enrollment and Officiation Object Values

✅ Closes: #124

* fix(domain/player): creates stubs for player's value objects

Co-authored-by: Yago Fernandes <yagoalmeida@gmail.com>

yfernandes committed:
Update issue templates

Commit:
refactor(valueobjects folder and its index.ts): valueObjects folder is updated with its VO and index

challenge.repository.interface.ts is now called IChallenge.repository.ts . VO were moved to
valueObjects folder in domain

Commit:
style(emailservice folder is no emailtemplates): email service stuff is implemented else where

email services are been implemented in notification.service.ts . Folder EmailService only contained
Email VO file and email templates. VO Email is now in valueObjects.ts and EmailService folder is now
called emailTemplates

Commit:
style(emailservice folder is no emailtemplates): email service stuff is implemented else where

email services are been implemented in notification.service.ts . Folder EmailService only contained
Email VO file and email templates. VO Email is now in valueObjects.ts and EmailService folder is now
called emailTemplates

angeliralves committed:
(ATQR-126) Create notifications enums (#72)

* feat(domain/service/notificationsenums.ts): add Notifications Enums

Add the file with the Notification Enums.

* feat(domain/services/notificationenums.ts): i fixed the errors

Fix the bugs pointed out.

* refactor(domain/notification): unifies notification related entities into services folder

Co-authored-by: Yago Fernandes <yagoalmeida@gmail.com>

Commit:
style(templates folder): templates folder has email folder with its tamplates

albnunes committed:
Merge branch 'development' into gabi/atqr-125-create-contacts-value-objects

albnunes committed:
fix: fix imports
